core/src/main/java/org/apache/iceberg/MetadataUpdate.java

```diff
@@ -223,31 +225,106 @@ public String name() {
 
     @Override
     public void applyTo(TableMetadata.Builder metadataBuilder) {
-      // TODO: this should be generalized when tagging is supported
-      metadataBuilder.removeBranch(name);
+      SnapshotRef ref = metadataBuilder.ref(name);
+      ValidationException.check(ref != null, "Ref %s not found", name);
+      if (ref.isBranch()) {
+        metadataBuilder.removeBranch(name);
+      } else {
+        metadataBuilder.removeTag(name);
+      }
     }
   }
 
   class SetSnapshotRef implements MetadataUpdate {
-    private final String name;
-    private final long snapshotId;
+    private final String ref;
+    private final Long snapshotId;
+    private final SnapshotRefType type;
+    private Integer minSnapshotsToKeep;
+    private Long maxSnapshotAgeMs;
+    private Long maxRefAgeMs;
+
+    public SetSnapshotRef(String ref, Long snapshotId, SnapshotRefType type, Integer minSnapshotsToKeep,
+                          Long maxSnapshotAgeMs, Long maxRefAgeMs) {
+      this.ref = ref;
+      this.snapshotId = snapshotId;
+      this.type = type;
+      this.minSnapshotsToKeep = minSnapshotsToKeep;
+      this.maxSnapshotAgeMs = maxSnapshotAgeMs;
+      this.maxRefAgeMs = maxRefAgeMs;
+    }
 
-    public SetSnapshotRef(String name, long snapshotId) {
-      this.name = name;
+    public SetSnapshotRef(String ref, Long snapshotId, SnapshotRefType type) {
+      this.ref = ref;
       this.snapshotId = snapshotId;
+      this.type = type;
     }
 
-    public String name() {
-      return name;
+    public String ref() {
+      return ref;
+    }
+
+    public SnapshotRefType type() {
+      return type;
     }
 
     public long snapshotId() {
       return snapshotId;
     }
 
+    public Integer minSnapshotsToKeep() {
+      return minSnapshotsToKeep;
+    }
+
+    public Long maxSnapshotAgeMs() {
+      return maxSnapshotAgeMs;
+    }
+
+    public Long maxRefAgeMs() {
+      return maxRefAgeMs;
+    }
+
     @Override
     public void applyTo(TableMetadata.Builder metadataBuilder) {
-      metadataBuilder.setBranchSnapshot(snapshotId, name);
+      SnapshotRef currentRef = metadataBuilder.ref(ref);
+      Preconditions.checkArgument(currentRef == null || currentRef.type() == type,
+          "Cannot apply SetSnapshotRef where current ref is of type %s and update is of type %s",
+          currentRef.type(), type);
+      SnapshotRef.Builder builder;
+      boolean newRef = currentRef == null;
+
+      if (!newRef) {
+        builder = SnapshotRef.builderFrom(currentRef);
+      } else {
+        builder = SnapshotRef.builderFor(snapshotId, type);
+      }
+
+      if (newRef || (currentRef.isBranch() && minSnapshotsToKeep != null)) {
```

Comment:
I think these should be set if `minSnapshotsToKeep` is set. If the ref is a branch, then the builder should reject this. And if this is a new ref, then it should only be set if it is non-default (non-null).